client/state/use-reducer.js

```diff
@@ -0,0 +1,54 @@
+/**
+ * External dependencies
+ */
+import { createElement, Component } from 'react';
+
+/**
+ * Internal dependencies
+ */
+import { addReducer, removeReducer } from 'state';
+
+/**
+ * Higher Order Component to add a reducer to the store's reducer list.
+ * TODO: Add this file to a place accessible by all extensions when it's ready.
+ *
+ * @param { string } reducerName The name for the reducer/key for redux state tree.
+ * @param { func } reducerFunc The function that takes (state, action) and returns state.
+ * @returns { func } A curried function that takes a component to wrap.
+ */
+export default function useReducer( reducerName, reducerFunc ) {
+	return function wrapWithUseReducer( WrappedComponent ) {
+		class UseReducer extends Component {
+			constructor( props, context ) {
+				super( props, context );
+				this.store = props.store || context.store;
+				this.reducerName = reducerName;
+				this.reducerFunc = reducerFunc;
+			}
+
+			componentDidMount() {
+				addReducer( reducerName, reducerFunc );
+			}
+
+			componentWillReceiveProps( newProps ) {
+				if ( newProps.store ) {
+					removeReducer( reducerName );
+					this.store = newProps.store;
```

Comment:
Oops, so looking over the code again, I see one good reason to keep this around, and it's to compare if the store was provided as a prop and has changed. Granted, that probably "should never happen" in our application, so maybe we can ignore that case. It's enough of an edge case that I'm comfortable removing support for it if we all agree.